cpp_wrapper: Precompile device-specific header files

benjaminglass1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #146452
* #146706
* #146424
* #146109
* #146449
* #144349
* #144293
* __->__ #144002

This saves us about a second per compilation, which is _massive_ for the OpInfo tests. Total OpInfo test runtime is down about 2x from this change alone.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D69185685](https://our.internmc.facebook.com/intern/diff/D69185685)

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

malfet commented:
> This saves us about a second per compilation, which is massive for the OpInfo tests. Total OpInfo test runtime is down about 2x from this change alone.

@benjaminglass1 would it be possible to add some sort of end-to-end script one can run to observe the build saving time?
Also, would be good to add some description on what measures are taken to invalidate the PCH when underlying headers change

benjaminglass1 commented:
@malfet I think I could come up with some kind of script, sure! WRT invalidation, the precompiled headers are lazily generated when required, and deleted when the program runtime ends, so that shouldn't be an issue. They get invalidated this strongly because precompiled headers are pretty strongly dependent on specific compiler versions and flags, so there's no point keeping them when recompiling is pretty quick.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
> > This saves us about a second per compilation, which is massive for the OpInfo tests. Total OpInfo test runtime is down about 2x from this change alone.
> 
> @benjaminglass1 would it be possible to add some sort of end-to-end script one can run to observe the build saving time?

@malfet This is what I came up with; on my system, it demonstrates a compile time drop from 2.73s to 1.45s on this PR. On the next commit in the stack, it demonstrates a further compile time drop to 0.90s.

```python
import statistics
import time

from torch._inductor.codecache import CppWrapperCodeCache
from torch._inductor.utils import fresh_inductor_cache

# Source code grabbed from a run of cpp_wrapper that generates a length-100 tensor of
# zeroes.  The precise code doesn't matter much; just that it's relatively simple (and
# should compile decently quickly).
SOURCE_CODE = """
#include <torch/csrc/inductor/cpp_wrapper/cpu.h>

#include <torch/csrc/inductor/cpp_prefix.h>
extern "C"  void cpp_fused_zeros_0(float* out_ptr0)
{
    {
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(100L); x0+=static_cast<int64_t>(8L))
        {
            {
                if(C10_LIKELY(x0 >= static_cast<int64_t>(0) && x0 < static_cast<int64_t>(96L)))
                {
                    auto tmp0 = static_cast<float>(0.0);
                    auto tmp1 = at::vec::Vectorized<float>(tmp0);
                    tmp1.store(out_ptr0 + static_cast<int64_t>(x0));
                }
                if(C10_UNLIKELY(x0 >= static_cast<int64_t>(96L) && x0 < static_cast<int64_t>(100L)))
                {
                    auto tmp0 = static_cast<float>(0.0);
                    auto tmp1 = at::vec::Vectorized<float>(tmp0);
                    tmp1.store(out_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(4L));
                }
            }
        }
    }
}
CACHE_TORCH_DTYPE(float32);
CACHE_TORCH_DEVICE(cpu);

void inductor_entry_impl(
    AtenTensorHandle*
        input_handles, // array of input AtenTensorHandle; handles
                        // are stolen; the array itself is borrowed
    AtenTensorHandle*
        output_handles  // array for writing output AtenTensorHandle; handles
                        // will be stolen by the caller; the array itself is
                        // borrowed)
) {
    py::gil_scoped_release release;

    auto inputs = steal_from_raw_handles_to_raii_handles(input_handles, 0);
    static constexpr int64_t int_array_0[] = {100L, };
    static constexpr int64_t int_array_1[] = {1L, };
    AtenTensorHandle buf0_handle;
    AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_empty_strided(1, int_array_0, int_array_1, cached_torch_dtype_float32, cached_torch_device_type_cpu,  0, &buf0_handle));
    RAIIAtenTensorHandle buf0(buf0_handle);
    cpp_fused_zeros_0((float*)(buf0.data_ptr()));
    output_handles[0] = buf0.release();
} // inductor_entry_impl

"""


def time_compile() -> float:
    with fresh_inductor_cache():
        start_time = time.perf_counter()
        CppWrapperCodeCache.load_pybinding(
            ["std::vector<AtenTensorHandle>"], SOURCE_CODE, "cpu", 1
        )
        end_time = time.perf_counter()
    return end_time - start_time


def average_time_compile() -> float:
    return statistics.median(time_compile() for _ in range(10))


print(f"With precompiled headers: {average_time_compile():0.2f}s")
CppWrapperCodeCache._get_uncompiled_header = classmethod(lambda cls, device: None)
print(f"Without: {average_time_compile():0.2f}s")
```

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
> @malfet I think I could come up with some kind of script, sure! WRT invalidation, the precompiled headers are lazily generated when required, and deleted when the program runtime ends, so that shouldn't be an issue. They get invalidated this strongly because precompiled headers are pretty strongly dependent on specific compiler versions and flags, so there's no point keeping them when recompiling is pretty quick.

@malfet I need to retract this statement, in part. When I ran the AOT Inductor HUD benchmarks, the compile times got about 3 seconds slower on average, leading me to believe that each benchmark was individually precompiling the headers. I've switched to using the same global cache dir as everything else in Inductor.

To handle invalidation properly, the precompiled header hash now contains (1) the header name, (2) the set of flags to be used when precompiling, and (3) the version string output from the compiler. These three should uniquely identify a precompiled header.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -1945,6 +1961,60 @@ def convert_arg(arg: Any) -> Any:
+) -> str:
+    compiler_version = get_compiler_version_info(compiler)
+    header_hash, header_full_path = write(
+        content=f"#include <{header}>\n",
```
I think it makes sense to hash on the actual header file content (as a transitive closure)?

benjaminglass1 replied:
@desertfire I considered this, but part of the issues is that the contents of headers further down the stack could change without affecting this top-level header, and we wouldn't appropriately rebuild in that case. I spent some time listing out scenarios that would invalidate a precompiled header, and I'm reasonably confident that all of them are covered by including the extra hashable content below.

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -1945,6 +1961,60 @@ def convert_arg(arg: Any) -> Any:
+        specified_dir=_HEADER_DIR,
+    )
+
+    if header_hash not in _COMPILED_HEADERS:
```
With proper hashing, I think we can just check if the precompiled header file exists here.

benjaminglass1 replied:
@desertfire Agreed that we could mostly do that at this point, but `_worker_compile_cpp` does that internally already and skips the compilation process if the file exists. I don't see any evidence locally that we lose much performance by leaving the file existence check in a single place here.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

desertfire commented on tools/generate_torch_version.py:
```diff
@@ -97,13 +110,19 @@ def get_torch_version(sha: str | None = None) -> str:
         f.write(f"__version__ = '{version}'\n")
+        # For builds using a git worktree with changes, add a guaranteed-unique build
+        # number that can be used to invalidate caches.
+        build_uuid = str(uuid4()) if git_worktree_dirty(pytorch_root) else None
```
Is it possible to add a test for this?

benjaminglass1 replied:
@desertfire Should be, I'll get on that.

desertfire replied:
I think you can mock git_worktree_dirty to force a recompilation, e.g. adding a test only option to https://github.com/pytorch/pytorch/blob/cd6c0707a86d77f9a21821b78904199a047b1bb0/torch/_inductor/config.py#L1476, and then check the recompilation does happen in a test.

benjaminglass1 replied:
I'll do this in a follow-up PR!

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -1648,14 +1650,28 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
+                **compile_command,  # type: ignore[arg-type]
             )
+            object_build_options.set_precompiled_header(
+                _precompile_header(
```
Let's separate this for cpp-wrapper and AOTI. I would like to see this rolling out for cpp-wrapper first, and then later applying to AOTI. Also it's worth to add an inductor config to control this behavior, in case caching goes bad and we can easily triage.

benjaminglass1 replied:
@desertfire Would you prefer separate commits, or would separate config variables do the trick?

benjaminglass1 replied:
I went with config variables, unless you have objections.

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -1944,6 +1960,60 @@ def convert_arg(arg: Any) -> Any:
+# specific compiler version and set of flags.  We explicitly use default_cache_dir here
+# because these headers need to be global, rather than ignored by fresh_inductor_cache.
+_HEADER_DIR = os.path.join(default_cache_dir(), "precompiled_headers")
+_COMPILED_HEADERS: OrderedSet[str] = OrderedSet()
```
This is not thread-safe. Since _worker_compile_cpp checks the file existence, we don't really this one?

benjaminglass1 replied:
@desertfire I'll run some tests locally to verify that there's no particular slowdown from always querying the filesystem, but you're probably correct.

benjaminglass1 replied:
@desertfire You were indeed correct. I've dropped this construct entirely.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
@desertfire I've stacked the remainder of the test bugfix stack underneath this, temporarily, in order to test whether the speedups here allow us to run the full suite of inductor tests. Hopefully we'll just be able to merge those out from underneath this, but if not I'll unstack whenever you need.

benjaminglass1 committed:
Update

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69185685).

desertfire reviewed (approved):
Chatted offline and test will come in a follow-up PR.

benjaminglass1 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but this breaks some inductor tests running internally' -c ghfirst

```
======================================================================
ERROR: test_cpp_wrapper (caffe2.test.inductor.test_cudagraph_trees.CudaGraphTreeTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/fbcode/platform010/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/caffe2/test/inductor/test_cudagraph_trees.py", line 2503, in test_cpp_wrapper
    compiled_result = self.run_twc(compiled, example_input)
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/caffe2/test/inductor/test_cudagraph_trees.py", line 197, in run_twc
    fn(*args, **kwargs)
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_dynamo/eval_frame.py", line 574, in _fn
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/compile_fx.py", line 752, in _compile_fx_inner
    raise InductorError(e, currentframe()).with_traceback(
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/compile_fx.py", line 737, in _compile_fx_inner
    mb_compiled_graph = fx_codegen_and_compile(
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/compile_fx.py", line 1402, in fx_codegen_and_compile
    return scheme.codegen_and_compile(gm, example_inputs, inputs_to_check, graph_kwargs)
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/compile_fx.py", line 1122, in codegen_and_compile
    compiled_fn = graph.compile_to_module().call
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/graph.py", line 1986, in compile_to_module
    return self._compile_to_module()
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/graph.py", line 2028, in _compile_to_module
    mod = PyCodeCache.load_by_key_path(
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/codecache.py", line 2853, in load_by_key_path
    mod = _reload_python_module(key, path)
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/runtime/compile_tasks.py", line 51, in _reload_python_module
    exec(code, mod.__dict__, mod.__dict__)
  File "/re_tmp/tmpwrjbuu4z/jq/cjq53uwok6fj5qand5kfwnazczljuwix4xupx7gmoxjlon2f4hea.py", line 197, in <module>
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/codecache.py", line 2347, in load_pybinding
    return cls.load_pybinding_async(*args, **kwargs)()
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/codecache.py", line 2328, in load_pybinding_async
    get_result = cls.load_async(
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/codecache.py", line 2096, in load_async
    _precompile_header(
  File "/re_cwd/buck-out/v2/gen/fbcode/bbdd369aded5cb42/caffe2/test/inductor/__cudagraph_trees__/cudagraph_trees#link-tree/torch/_inductor/codecache.py", line 1974, in _precompile_header
    extra=f"{hashable_cmd_line} {compiler_version} {torch.version.__version__} {torch.version.build_uuid}",
torch._inductor.exc.InductorError: AttributeError: module 'torch.version' has no attribute 'build_uuid'

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


----------------------------------------------------------------------
Ran 1 test in 4.217s

FAILED (errors=1)
```

I think @desertfire has been looking into the failure

benjaminglass1 commented:
@desertfire @huydhn I'll open a new PR to re-land this in, to avoid any issues with the associated internal revision.
